src/xenia/kernel/xboxkrnl/xboxkrnl_memory.cc

```diff
@@ -495,33 +495,24 @@ dword_result_t MmQueryStatistics(
   stats_ptr->total_physical_pages = 0x00020000;  // 512mb / 4kb pages
   stats_ptr->kernel_pages = 0x00000300;
 
-  // TODO(gibbed): maybe use LookupHeapByType instead?
-  auto heap_a = kernel_memory()->LookupHeap(0xA0000000);
-  auto heap_c = kernel_memory()->LookupHeap(0xC0000000);
-  auto heap_e = kernel_memory()->LookupHeap(0xE0000000);
-
-  assert_not_null(heap_a);
-  assert_not_null(heap_c);
-  assert_not_null(heap_e);
-
-#define GET_USED_PAGE_COUNT(x) \
-  (x->GetTotalPageCount() - x->GetUnreservedPageCount())
-#define GET_USED_PAGE_SIZE(x) ((GET_USED_PAGE_COUNT(x) * x->page_size()) / 4096)
+  constexpr uint32_t heap_page_sizes[3] = {0x1000, 0x10000, 0x1000000};
   uint32_t used_pages = 0;
-  used_pages += GET_USED_PAGE_SIZE(heap_a);
-  used_pages += GET_USED_PAGE_SIZE(heap_c);
-  used_pages += GET_USED_PAGE_SIZE(heap_e);
-#undef GET_USED_PAGE_SIZE
-#undef GET_USED_PAGE_COUNT
+  uint32_t reserved_pages_bytes = 0;
 
-  assert_true(used_pages < stats_ptr->total_physical_pages);
+  for (uint8_t i = 0; i < std::size(heap_page_sizes); i++) {
+    xe::BaseHeap* heap =
+        kernel_memory()->LookupHeapByType(true, heap_page_sizes[i]);
+    assert_not_null(heap);
+
+    used_pages += (heap->GetUsedPagesCount() * heap->page_size()) / 4096;
```

Comment:
Here the global critical region lock is taken twice, not once, so there may be an allocation between `GetUsedPagesCount()` and `GetReservedPageCount()`, and in this case they may be inconsistent. But overall, you don't even need separate GetReserved and GetUnreserved, as they are (total minus each other).